### ISSUE TITLE
Write permission fault by ldelf

### ISSUE THREAD

**Issue by awerv** [OP-TEE/optee_os#4386]
Hi,

I'm trying to port a TA to iMX6 which I developed in the QEMU based environment. Since there is no manifest file for iMX, I'm using Buildroot to compile everything, including `optee_os`. I'm using version 3.6.0 of `optee_os`, `optee_client` and `optee_test`.
OPTEE gives the following logs:

```
D/TC:? 0 tee_ta_init_pseudo_ta_session:280 Lookup pseudo TA c6ccbdc9-0561-4fe5-8eec-e7e2f1deaa05
D/TC:? 0 load_ldelf:744 ldelf load address 0x104000
D/LD:  ldelf:117 Loading TA c6ccbdc9-0561-4fe5-8eec-e7e2f1deaa05
D/TC:? 0 tee_ta_init_session_with_context:570 Re-open TA 3a2f8978-5dc0-11e8-9c2d-fa7ae01bbebc
D/TC:? 0 system_open_ta_binary:229 Lookup user TA ELF c6ccbdc9-0561-4fe5-8eec-e7e2f1deaa05 (Secure Storage TA)
D/TC:? 0 system_open_ta_binary:232 res=0xffff0008
D/TC:? 0 system_open_ta_binary:229 Lookup user TA ELF c6ccbdc9-0561-4fe5-8eec-e7e2f1deaa05 (REE [buffered])
D/TC:? 0 system_open_ta_binary:232 res=0x0
D/TC:0 0 abort_handler:520 [abort] abort in User mode (TA will panic)
E/TC:? 0 
E/TC:? 0 User TA data-abort at address 0x11adb4 (write permission fault)
E/TC:? 0  fsr 0x0000080f  ttbr0 0x4e07e04a  ttbr1 0x4e07404a  cidr 0x2
E/TC:? 0  cpu #0          cpsr 0x60000030
E/TC:? 0  r0 0x00000017      r4 0x00000003    r8 0x00000001   r12 0x0011adb4
E/TC:? 0  r1 0x0010d512      r5 0x0010bf70    r9 0x0011b6bc    sp 0x00103f70
E/TC:? 0  r2 0x0000ddb4      r6 0x0011b68c   r10 0x00000016    lr 0x00000017
E/TC:? 0  r3 0x0010d000      r7 0x0011c0ec   r11 0x00000028    pc 0x00105bb2
E/TC:? 0  region  0: va 0x00100000 pa 0x4e000000 size 0x001000 flags ---R-X
E/TC:? 0  region  1: va 0x00102000 pa 0x4e100000 size 0x002000 flags rw-RW-
E/TC:? 0  region  2: va 0x00104000 pa 0x4e102000 size 0x005000 flags r-x---
E/TC:? 0  region  3: va 0x00109000 pa 0x4e107000 size 0x001000 flags rw-RW-
E/TC:? 0  region  4: va 0x0010a000 pa 0x4e108000 size 0x002000 flags rw-RW-
E/TC:? 0  region  5: va 0x0010c000 pa 0x4e11b000 size 0x001000 flags r--R--
E/TC:? 0  region  6: va 0x0010d000 pa 0x4e11c000 size 0x00f000 flags r-xR--
E/TC:? 0  region  7: va 0x0011c000 pa 0x4e12b000 size 0x00c000 flags rw-RW-
E/TC:? 0  region  8: va 0x00128000 pa 0x4e10a000 size 0x001000 flags rw-RW-
E/TC:? 0 init_with_ldelf:275 ldelf panicked
D/TC:? 0 tee_ta_close_session:496 csess 0x4e071788 id 1
D/TC:? 0 tee_ta_close_session:515 Destroy session
D/TC:? 0 destroy_context:296 Destroy TA ctx (0x4e071748)
D/TC:? 0 tee_ta_close_session:496 csess 0x4e071328 id 1
D/TC:? 0 tee_ta_close_session:515 Destroy session
E/TC:? 0 tee_ta_open_session:725 Failed. Return error 0xffff0000
```

I was looking for answers among other issues and first found #1355, but I'm not using any parameters when invoking the TA.

Next I found #3825. My TA uses libraries only compiled by `optee_os`, namely `libutils.a`, `libutee.a` and `libmbedtls.a`. In these archives I found `.o` files with `.rel.text` sections. I also checked the archives compiled by the QEMU environment, where readelf didn't find any `.rel.text` sections, so I guess this is the root of my issue.

Since it was suggested in #3825, I was trying to get rid of these sections, but I didn't succeed. I tried the following patches:

```diff
diff --git a/mk/lib.mk b/mk/lib.mk
index 007a02d1..a9813692 100644
--- a/mk/lib.mk
+++ b/mk/lib.mk
@@ -21,7 +21,7 @@ subdirs = $(libdir)
 include mk/subdir.mk
 ifeq ($(filter $(sm), core ldelf),) # TA
 ifeq ($(CFG_ULIBS_MCOUNT),y)
-cflags-lib$(libname)-$(sm) += -pg
+cflags-lib$(libname)-$(sm) += -pg -fPIC
 endif
 endif
 include mk/compile.mk
```

```diff
diff --git a/mk/compile.mk b/mk/compile.mk
index 6a2336e1..1644a07c 100644
--- a/mk/compile.mk
+++ b/mk/compile.mk
@@ -17,7 +17,7 @@ objs          :=
 # Disable all builtin rules
 .SUFFIXES:
 
-comp-cflags$(sm) = -std=gnu99
+comp-cflags$(sm) = -std=gnu99 -fPIC
 comp-aflags$(sm) =
 comp-cppflags$(sm) =

```

```diff
diff --git a/lib/libmbedtls/sub.mk b/lib/libmbedtls/sub.mk
index 66d61b74..63f1ca2d 100644
--- a/lib/libmbedtls/sub.mk
+++ b/lib/libmbedtls/sub.mk
@@ -94,6 +94,7 @@ srcs-$(sm-$(ta-target)) += $(addprefix mbedtls/library/, $(SRCS_TLS))
 
 cflags-lib-y += -Wno-redundant-decls
 cflags-lib-y += -Wno-switch-default
+cflags-lib-y += -fPIC
 
 ifeq ($(CFG_CRYPTOLIB_NAME_mbedtls),y)
 subdirs-$(sm-core) += core
diff --git a/lib/libmpa/sub.mk b/lib/libmpa/sub.mk
index 31d8295b..f51df462 100644
--- a/lib/libmpa/sub.mk
+++ b/lib/libmpa/sub.mk
@@ -37,4 +37,6 @@ srcs-y += mpa_init.c
 srcs-y += mpa_io.c
 srcs-y += mpa_modulus.c
 
+cflags-lib-y += -fPIC
 subdirs-$(arch_arm) += arch/$(ARCH)
diff --git a/lib/libutee/sub.mk b/lib/libutee/sub.mk
index 7f3b3b20..716f62a3 100644
--- a/lib/libutee/sub.mk
+++ b/lib/libutee/sub.mk
@@ -22,4 +22,6 @@ srcs-y += tee_api_arith_mpa.c
 endif
 endif #ifneq ($(sm),ldelf)
 
+cflags-lib-y += -fPIC
 subdirs-y += arch/$(ARCH)
diff --git a/lib/libutils/sub.mk b/lib/libutils/sub.mk
index e49bd899..5250564d 100644
--- a/lib/libutils/sub.mk
+++ b/lib/libutils/sub.mk
@@ -1,2 +1,4 @@
 subdirs-$(CFG_LIBUTILS_WITH_ISOC) += isoc
 subdirs-y += ext
+cflags-lib-y += -fPIC

```

Unfortunately, none of these worked, I still have `.rel.text` segments in the archives. Moreover, Buildroot config states it will compile everything to be position independent: `BR2_PIC_PIE=y`.

I also tried to make Buildroot use an external toolchain. I gave it the one used by the QEMU environment, but it didn't work, Buildroot says it's incompatible (I guess because it is for EABIhf, while the only available target ABI in my Buildroot project is EABI). Currently I'm using the toolchain provided by Buildroot.

The funny part is that every TA of the `optee_test` project works fine. I guess they are not using any function from the problematic sections.

Bottom line is, I'm out of ideas about how to get rid of those sections. Can anyone please advise? Thanks in advance.

**Comment by jforissier** [comment]
Hi @awerv,

Thanks for the detailed description of the issue. Indeed as long as you have "text relocations", things will not work. We need to understand where these relocations are coming from. For this I think I have a small script that could be quite helpful :) See https://github.com/jforissier/misc/blob/master/print_relocations.py.
Just give it your TA ELF file and check for relocations inside `.text`:
```
$ ./print_relocations.py ta_file.elf | grep '\.text'
```
It will show the addresses of the relocation targets and translate them into "section <symbol + offset>".

**Comment by awerv** [comment]
Hi @jforissier,

Thanks for the script! Grep exits with 1, meaning it didn't find anything to match. I checked it, the script only found relocations in the `.data`, `.got` and `.pv_table` sections of the TA ELF. Also readelf doesn't report any `.rel.text` section from the TA ELF.

I also tried your script against one of the mentioned archives, but it fails, since it's not an ELF file.

**Comment by jforissier** [comment]
Weird. Perhaps the faulting address will help? 0x11adb4 - 0x0010d000 = 0xddb4, look at this address in the TA ELF (in the `<uuid>.dmp` file for example).

**Comment by awerv** [comment]
That's where I found it in the `dmp` file:

```
SYMBOL TABLE:
00000000 l    d  .ta_head       00000000 .ta_head
00000020 l    d  .text          00000000 .text
0000c4d0 l    d  .rodata        00000000 .rodata
0000dcb8 l    d  .ARM.extab     00000000 .ARM.extab
0000dd9c l    d  .gnu.hash      00000000 .gnu.hash
0000ddb4 l    d  .pv_table      00000000 .pv_table
etc...
```

**Comment by jforissier** [comment]
Well if you can't figure out what is in you TA at address 0xddb4, there are little chance I can.

**Comment by jforissier** [comment]
You mentioned you are using OP-TEE 3.6.0. Could you try latest? The issue might have been fixed since.

**Comment by awerv** [comment]
I managed to get rid of `.pv_table` and the issue is gone. Thank you for pointing me to the right direction, I'm closing the issue.
